sdk/core/core-https/src/policies/proxyPolicy.ts

```diff
@@ -7,28 +7,78 @@ import { URL } from "../util/url";
 
 const HTTPS_PROXY = "HTTPS_PROXY";
 const HTTP_PROXY = "HTTP_PROXY";
+const ALL_PROXY = "ALL_PROXY";
+const NO_PROXY = "NO_PROXY";
 
 /**
  * The programmatic identifier of the proxyPolicy.
  */
 export const proxyPolicyName = "proxyPolicy";
 
+export const noProxyList: string[] = loadNoProxy();
+const byPassedList: Map<string, boolean> = new Map();
+
+function getEnvironmentValue(name: string): string | undefined {
+  if (process.env[name]) {
+    return process.env[name];
+  } else if (process.env[name.toLowerCase()]) {
+    return process.env[name.toLowerCase()];
+  }
+  return undefined;
+}
+
 function loadEnvironmentProxyValue(): string | undefined {
   if (!process) {
     return undefined;
   }
 
-  if (process.env[HTTPS_PROXY]) {
-    return process.env[HTTPS_PROXY];
-  } else if (process.env[HTTPS_PROXY.toLowerCase()]) {
-    return process.env[HTTPS_PROXY.toLowerCase()];
-  } else if (process.env[HTTP_PROXY]) {
-    return process.env[HTTP_PROXY];
-  } else if (process.env[HTTP_PROXY.toLowerCase()]) {
-    return process.env[HTTP_PROXY.toLowerCase()];
+  const httpsProxy = getEnvironmentValue(HTTPS_PROXY);
+  const allProxy = getEnvironmentValue(ALL_PROXY);
+  const httpProxy = getEnvironmentValue(HTTP_PROXY);
+
+  return httpsProxy || allProxy || httpProxy;
+}
+
+// Check whether the host of a given `uri` is in the noProxyList.
+// If there's a match, any request sent to the same host won't have the proxy settings set.
+function isBypassed(uri: string): boolean | undefined {
+  if (noProxyList.length === 0) {
+    return false;
+  }
+  const host = new URL(uri).hostname;
+  if (byPassedList.has(host)) {
+    return byPassedList.get(host);
   }
+  let isBypassedFlag = false;
+  for (const proxyString of noProxyList) {
+    if (proxyString[0] === ".") {
+      if (uri.endsWith(proxyString)) {
+        isBypassedFlag = true;
+      } else {
+        if (host === proxyString.slice(1) && host.length === proxyString.length - 1) {
```

Comment:
To check my understanding, this if for if the proxyString is `.microsoft.com` and the host is `microsoft.com` then it'll return true?

Reply:
Good question!. I did some Internet search study. Some findings:
- NO_PROXY is more a convention than a standard so there isn't a spec for it.
- in some libraries/programs, host name of `example.com` would NOT match the pattern of `.example.com` (requiring subdomain to match), while in others, e.g., `curl` it would.
- We probably don't support all of the ignore patterns properly. Other things include: a single `*` character, CIDR addresses, port.

I'll log a separate issue for looking more into the ignore patterns.

Reply:
well, I think it's a direct port from .NET Azure.Core https://github.com/Azure/azure-sdk-for-net/blob/8cca811371159e527159c7eb65602477898683e2/sdk/core/Azure.Core/src/Pipeline/Internal/HttpEnvironmentProxy.cs#L210. However, there's some difference. For example, .NET uses `input.Host.EndsWith` while we have `uri.endsWith`. .NET has length check first likely to short-circuit the string comparison.

I will update our code to match .NET's in this PR.

Reply:
I believe `proxyString.slice(1) ` is to remove the leading `*` from bypass pattern.